File: Practices(youknowh0)/Recursion/Apple_Division.cpp
#include <bits/stdc++.h> 

using namespace std;

long long totalSUM = 0;
long long minDiff = LLONG_MAX;

void divide(int index, long long group1Sum, const vector <int> &v){
    if(index == v.size()){
        long long group2Sum = totalSUM - group1Sum;
        minDiff = min(minDiff, abs(group1Sum - group2Sum));
        return;
    }

    divide(index + 1, group1Sum + v[index], v);
    divide(index + 1, group1Sum, v);
}

int main(){
    int n;
    cin >> n;

    vector <int> v(n);
    for(auto &i:v){
        cin >> i;
        totalSUM += i;
    }

    divide(0,0,v);
    cout << minDiff << endl;
    return 0;
}

File: Practices(youknowh0)/Recursion/Apple_division.py
def divide(index, group1_sum, weights, total_sum, path, depth=0):
    indent = "  " * depth
    group2_sum = total_sum - group1_sum
    diff = abs(group1_sum - group2_sum)
    
    # Print current recursive state
    print(f"{indent}Index {index}: Group1 = {group1_sum}, Group2 = {group2_sum}, Diff = {diff}, Path: {path}")
    
    # Base case
    if index == len(weights):
        return

    # Option 1: Include current weight in group1
    divide(index + 1, group1_sum + weights[index], weights, total_sum, path + [weights[index]], depth + 1)
    
    # Option 2: Exclude current weight (goes to group2)
    divide(index + 1, group1_sum, weights, total_sum, path, depth + 1)

# Example input
weights = [3, 2, 7, 4, 1]
total_sum = sum(weights)

print("Recursive Call Tree:\n")
divide(0, 0, weights, total_sum, [])
